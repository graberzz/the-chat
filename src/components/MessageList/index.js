import React from 'react';
import { CSSTransitionGroup } from 'react-transition-group';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Message from '../Message';

const MessageListWrap = styled.ul`
  height: 100%;
  overflow-y: scroll;
  overflow-x: hidden;

  &::after {
    content: '';
    display: table;
    clear: both;
  }
`;

class MessageList extends React.Component {  
  componentDidMount() {
    this.wrap.scrollTo(0, this.wrap.scrollHeight);
  }
  
  componentDidUpdate() {
    this.wrap.scrollTo(0, this.wrap.scrollHeight);
  }
  
  render() {
    const { messages, user } = this.props;

    return (
      <MessageListWrap innerRef={ref => this.wrap = ref}>
        <CSSTransitionGroup
          transitionName='slidein'
          transitionEnterTimeout={500}
          transitionLeaveTimeout={500}>
          {messages.map((msg) =>
            <Message
              key={msg.id}
              own={user && msg.user.email === user.email}
              avatar={msg.user.avatar}
              username={msg.user.username}
              message={msg.message}
              time={new Date(msg.timestamp).toLocaleString('ru-RU')}
            />
          )}
        </CSSTransitionGroup>
      </MessageListWrap>
    );
  }
}

MessageList.propTypes = {
  messages: PropTypes.arrayOf(PropTypes.shape({
    message: PropTypes.string.isRequired,
    user: PropTypes.shape({
      email: PropTypes.string.isRequired,
    }),
  })).isRequired,
  user: PropTypes.shape({
    email: PropTypes.string.isRequired,
  }),
};

export default MessageList;