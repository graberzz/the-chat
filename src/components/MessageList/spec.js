import React from 'react';
import { render } from 'enzyme';
import MessageList from './';

const testMsgs = [
  {
    id: '1',
    message: 'msg1',
    timestamp: 1,
    user: {
      email: '@mail',
      username: 'test',
      avatar: 'avatar'
    },
  },
  {
    id: '2',
    message: 'msg2',
    timestamp: 3,
    user: {
      email: '@mail',
      username: 'test',
      avatar: 'avatar'
    },
  },
];

describe('<MessageList />', () => {
  it('matches snapshot', () => {
    expect(render(<MessageList messages={testMsgs} timestampToDate={ts => ts} />)).toMatchSnapshot();
  });
});