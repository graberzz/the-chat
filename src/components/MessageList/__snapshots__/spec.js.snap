// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList /> matches snapshot 1`] = `
<ul
  class="sc-iwsKbI biXbNh"
>
  <span>
    <li
      class="sc-ifAKCX eJhyaS"
    >
      <div
        class="sc-bZQynM ifsTat"
      >
        <div
          class="sc-gzVnrw gczGLK"
        >
          test
        </div>
        <img
          alt="test"
          class="sc-htoDjs ftOaTd"
          src="avatar"
        >
          
        </img>
      </div>
      msg1
      <div
        class="sc-dnqmqq iDZheP"
      >
        1970-1-1 05:00:00
      </div>
    </li>
    <li
      class="sc-ifAKCX eJhyaS"
    >
      <div
        class="sc-bZQynM ifsTat"
      >
        <div
          class="sc-gzVnrw gczGLK"
        >
          test
        </div>
        <img
          alt="test"
          class="sc-htoDjs ftOaTd"
          src="avatar"
        >
          
        </img>
      </div>
      msg2
      <div
        class="sc-dnqmqq iDZheP"
      >
        1970-1-1 05:00:00
      </div>
    </li>
  </span>
</ul>
`;
